### ISSUE TITLE
ProtocolMaxMsgSize vs block max size

### ISSUE THREAD

**Issue by pnomarev** [ethereum/go-ethereum#2152]
There is a potential issue: theoretically block size is unlimited, limited only by GasLimit which can grow over time.
Therefore generated block size may become larger than the maximum allowed protocol message size, which will lead to the node isolation

**Comment by fjl** [comment]
I agree that this is a theoretical issue, and thank you for noticing.

However, consider the following:

The largest message size allowed for eth is 10MB at this time. It applies to all messages,
even those containing multiple blocks or transactions. The code that sends batches of blocks tries
to keep the message size below 2MB (`softResponseLimit`), the only case where this limit could
be exceeded is if a single block or transaction is larger than 2MB.

The largest block to date was 2kB. We're still operating with pi million gas limit though,
so when it raises itself blocks may become larger. But there is only so much execution
that can fit with 15s block time. If someone generates a 4MB 'jumbo' block, it will likely
take a long time to execute and a long time to propagate. So there is a bit of an incentive
not to make such large blocks.

**Comment by pnomarev** [comment]
This issue is easy to reproduce on a private blockchain. I'm not sure whether this whole project is for public network only, or it has to provide a reliable private network as well.
Currently this "extreme case" issue prevent it from being used in a private environment, where difficulty and GasLimit can be adjusted and transactions load can be sometimes extremely high

**Comment by fjl** [comment]
Can you provide some details on the parameters you use for the private chain?

**Comment by pnomarev** [comment]
genesis:
{
	"nonce": "0xbebededefefeafdeef",
	"timestamp": "0x0",
	"parentHash": "0x0000000000000000000000000000000000000000000000000000000000000000",
	"extraData": "0x0",
	"gasLimit": "0x80000000",
	"difficulty": "0x100000",
	"mixhash": "0x0000000000000000000000000000000000000000000000000000000000000000",
	"coinbase": "0x3333333333333333333333333333333333333333"
}

- running two nodes on the same machine, connected to each other.
- first node is a miner
- mining speed without any transactions is roughly 1 block per 2-5 seconds
- submitting 1000 transactions/sec between 50 different accounts
- after ~20 blocks, the client would generate >10mb block and isolate itself(unable to propagate it to the rest of the network) from the second node due this issue.
